### DISCUSSION TITLE
Root loader data not available in the root ErrorBoundary

### DISCUSSION THREAD

**Discussion by meglio** [remix-run/remix#9043] [Q&A] [answered]
Any reason why root loader data is not available in the root ErrorBoundary?

I need global application data, that I usually supply through the root loader, to be available in both normal application rendering and in the root ErrorBoundary. For instance, if the path has not been found, the root loader still makes sense, doesn't it? But for some reason it is null.

**Comment by meglio** [comment]
The way it works now is: I read some data from a local config file, and return it as the root loader data. However, if a 404 Not Found error happens, I can't access that root loader data. For instance, I need to to read the application name - literally the name that is used in `<title>` and is written at the top of the layout.

**Comment by sergiodxa** [comment]
Because the reason why the root error boundary was rendered could be that the root loader failed, in that case there’s no root loader data.

You can use useRouteLoaderData to access it and handle the case when it’s undefined https://sergiodxa.com/tutorials/access-remix-s-loader-data-from-a-root-errorboundary

**Reply by meglio** [reply]
But the root loader has not failed, so why there is no root loader data available to the root error boundary? I don't get it.

**Reply by meglio** [reply]
Also, looks like `useRouteLoaderData` uses the same data that `useLoaderData()` does, while the latter is not allowed in ErrorBoundary - so how/why does it work?

**Reply by kiliman** [reply]
I believe they recommend `useRouteLoaderData` because of the types. `useLoaderData` assumes that the data will be available, whereas `useRouteLoaderData` does not (it returns `T | undefined`)

**Reply by sergiodxa** [reply]
How would ErrorBoundary know that it was not rendered because the root loader failed? That's why `useLoaderData` is not allowed there.

The `useRouteLoaderData` hook works because it considers the case when there's no loader data.

**Reply by meglio** [reply]
> How would ErrorBoundary know that it was not rendered because the root loader failed? 

Not sure what you mean by saying "how would it know" - it is programmed to do whatever we want it to do, no? So it could be a simple `if/else` logic to make it know, and the data could be either some non-empty data (if the loader have not failed) or null (if it failed). Same as `useRouteLoaderData()`, I suppose, - i.e. if data is there, it is available to the caller by the route ID; if it is not there, then it is not available. Why would `useLoaderData()` be not able to find it out when called inside the `ErrorBoundary`?

**Reply by meglio** [reply]
@sergiodxa so `useRouteLoaderData('root')` returns `undefined` in the root `ErrorBoundary`. Any other ideas?

**Reply by sergiodxa** [reply]
If it returns undefined it means the loader failed, are you throwing a response?

**Reply by meglio** [reply]
<img width="851" alt="image" src="https://github.com/remix-run/remix/assets/282177/32ea3eec-657e-4d27-947a-53df49552119">

I think the documentation is also a bit self-delusional in that it is not always the case when the loader failure is what led to ErrorBoundary. The "assumption" behind making loader data unavailable is wrong in the first place: it could have just returned "undefined" if the loader has failed, or the actual loader data if it has not. Do you see any reasonable explanation why `useLoaderData()` can't make that check but `useRouteLoaderData()` can?

**Reply by meglio** [reply]
> If it returns undefined it means the loader failed, are you throwing a response?

My root loader:

```
export function loader() {
    let data = get_global_application_data_from_config();
    return data;
}
```

This basically loads data from `config.ini`. If it does not fail on an existing route, I don't expect it to fail on a non-existent 404 route. Is my logic flawed?

**Reply by meglio** [reply]
Is root loader called for non-existent routes?

**Reply by kiliman** [reply]
It's not really "making a check". Those hooks simply return the values stored in Remix context.

Remix assumes that if you call `useLoaderData` then everything is on the _happy path_ and that you should always have a valid value.

With `useRouteLoaderData`, you can call it for any route id, even non-existent ones, so it is possible it will return `undefined`, since that route data won't be in the context.

Remix isn't being lazy here. It's very possible that for a given route that has, let's say 3 loaders, that all but 2 of them are successful, but the third throws an error. This will force Remix to render the nearest `ErrorBoundary`. At this point, it doesn't really know what loaders were successful or not. That's why it says, don't call `useLoaderData` since that assumes the _happy path_. But `useRouteLoaderData` will return whatever loaders were successful, and it's your responsibility to check which ones.

If you try to access a non-existent route, Remix will throw a _Not Found_ response, so it's possible that **none** of your loaders are called. So your loader didn't even get a chance to fail. Again, Remix can't guarantee the state of your loaders once you are in the `ErrorBoundary`.

Anyway, given that, my example will show how you can setup some app global data that doesn't depend on the root loader. I'm almost finished with it.

**Reply by meglio** [reply]
> Remix isn't being lazy here. It's very possible that for a given route that has, let's say 3 loaders, that all but 2 of them are successful, but the third throws an error. This will force Remix to render the nearest ErrorBoundary. At this point, it doesn't really know what loaders were successful or not. That's why it says, don't call useLoaderData since that assumes the happy path. But useRouteLoaderData will return whatever loaders were successful, and it's your responsibility to check which ones.

> If you try to access a non-existent route, Remix will throw a Not Found response, so it's possible that none of your loaders are called. So your loader didn't even get a chance to fail. Again, Remix can't guarantee the state of your loaders once you are in the ErrorBoundary.

Many thanks, these two are very thorough explanations, I'd put them into the Remix documentation as they clarify a lot of things.

**Comment by kiliman** [comment]
I've got a solution that lets you set global app data that can be used in any route, as well as the error boundary.

Request-specific data (like host header) will need to be returned from loaders, though. However, I also have a solution for returning request data to your error boundary as well.

I'll be posting an example repo later today. My example uses Vite, but I imagine it will work on the classic compiler.

**Reply by meglio** [reply]
Could you please post a link to that example here as well. Many thanks.
P.S. I use the latest Remix and in the Vite mode.

**Reply by meglio** [reply]
Note: I can see [define](https://vitejs.dev/config/shared-options.html#define) in Vite, but it does not solve the problem because:

> Entries will be defined as globals during dev and statically replaced during build.

I need it to pick up the data from the file when the server starts, and not hardcode it at the build stage.

**Reply by kiliman** [reply]
I replied to another of your issues with more detail. I will be posting the final example to GitHub.

**Comment by meglio** [comment]
I was also thinking about the [layout export](https://remix.run/docs/en/main/file-conventions/root#layout-export), e.g. to load the data from the file in there and expose it as a global context. But it is rendered on both server side and client side, so doesn't look like a solution either.

**Reply by meglio** [reply]
Can't we just augment the html returned by Layout with some data? like e.g. a `<script>` that sets a global variable. I suspect there will be cases when the Layout is, again, called on the client side, thus not being able to read from a file.

In general, I feel frustrated that a simple task of exposing some server-side data to the client side takes so much effort and thought. This is not normal, and it feels like Remix is standing in my way rather than helping me to do the job.

**Reply by kiliman** [reply, marked as answer]
I finally got my example finished and deployed. Check it out here.

https://github.com/kiliman/remix-global-data

There's even a link to a deployed app.

**Reply by meglio** [reply]
This is quite detailed, thanks @kiliman .
1. Should it work without typescript type declarations?
2. Another concern of mines is that it looks like I'll need to introduce`entry.server.tsx`? So what happens if it changes - will we now have to constantly watch for the changes in that entry with every release, and notice the change, and somehow update it when it gets changed?

This looks like a workaround that is not proper and is asking for problems in the future. It could have beeen a "global app data API" provided by Remix. Something in line with: we just export `getGlobalAppData` in `root.tsx` and that is it. Will you consider adding such an API?

**Reply by meglio** [reply]
One more thing. Tried it in `meta()` with the intention to vary the favicon depending on the app globals, and it did not work. So then I moved `<AppGlobalsScript />` into `<head>`  and put it just before `<Meta />` and `<Links />`, and it worked. Might be something to account for in your demo?

**Reply by kiliman** [reply]
True. I guess I don't see anything wrong with adding it to `<head>`. Good catch.

**Reply by kiliman** [reply]
> This looks like a workaround that is not proper and is asking for problems in the future. It could have beeen a "global app data API" provided by Remix. Something in line with: we just export getGlobalAppData in root.tsx and that is it. Will you consider adding such an API?

I can only give you a solution/workaround that would help you solve your problem today. If you think that is not "proper", there's not much I can do for you then. You can certainly wait for an official API, but you may be waiting for a long time. 

Just because you're using a framework doesn't mean you're not responsible for your own code. Sometimes, you'll have to write code that bridges the borders between your app and the framework. You can't expect a framework to support every use case. Luckily, Remix provides enough seams to let you add the necessary integration.

Personally, I don't have problems even using `patch-package` if it means getting my app to work the way I want it to.

**Reply by meglio** [reply]
> I can only give you a solution/workaround that would help you solve your problem today. If you think that is not "proper", there's not much I can do for you then. You can certainly wait for an official API, but you may be waiting for a long time.

I adopted the solution and it works. Thank you very much.

By naming it not proper, I didn't try to be unkind, sorry if it sounded so; just meant it probably should be considered a temporary workaround, and the idea to consider it to become some out-of-the-box API should not be abandoned.

> You can't expect a framework to support every use case.

Absolutely agree. It's just that the "global data" is not an unusual or weird way of using data in applications, hence suggesting it could be an API. I mean, with the general attitude of the Remix team to solve the hard problems for the user, it would fit well :) If it required a whole separate demo project with a readme to explain how it works, and also requried ejecting the server entry point file, I'd be happy to have a black box that "just works", which is how many other things are in Remix.

Anyway, thank you again for your prompt replies and the detailed demo project!

**Reply by kiliman** [reply]
Thanks. I agree that having this as an official API would be nice. Sometimes, it helps the team if there's a working example of implementing the API. So, hopefully, this can be used as a POC for future consideration.

**Reply by artemis-prime** [reply]
Yes please.  And with all due respect, no this is not "proper" and you shouldn't fault him for mentioning that.  It introduces significant complexity to fulfill a relatively common need in many apps.  Therefore a stable and clear solution clearly belongs in the framework!  Come on people!
